memstore/bootstrap.go

```diff
@@ -0,0 +1,479 @@
+//  Copyright (c) 2017-2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package memstore
+
+import (
+	"context"
+	m3Shard "github.com/m3db/m3/src/cluster/shard"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	xretry "github.com/m3db/m3/src/x/retry"
+	xsync "github.com/m3db/m3/src/x/sync"
+	"github.com/uber/aresdb/cluster/topology"
+	"github.com/uber/aresdb/datanode/bootstrap"
+	"github.com/uber/aresdb/datanode/client"
+	"github.com/uber/aresdb/datanode/generated/proto/rpc"
+	"github.com/uber/aresdb/utils"
+	"io"
+	"math"
+	"math/rand"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// WaitForMetaDataBootstrap waits for metadata bootstrap to be finished before certain process can proceed
+// such as recovery, ingestion
+func (shard *TableShard) WaitForMetaDataBootstrap() {
+	shard.bootstrapLock.Lock()
+	for shard.bootstrapState < bootstrap.MetaDataBootstrapped {
+		shard.readyForRecovery.Wait()
+	}
+	shard.bootstrapLock.Unlock()
+}
+
+// IsBootstrapped returns whether this table shard is already bootstrapped.
+func (shard *TableShard) IsBootstrapped() bool {
+	return shard.BootstrapState() == bootstrap.Bootstrapped
+}
+
+// BootstrapState returns this table shards' bootstrap state.
+func (shard *TableShard) BootstrapState() bootstrap.BootstrapState {
+	shard.bootstrapLock.RLock()
+	bs := shard.bootstrapState
+	shard.bootstrapLock.RUnlock()
+	return bs
+}
+
+// Bootstrap executes bootstrap for table shard
+func (shard *TableShard) Bootstrap(
+	peerSource client.PeerSource,
+	origin topology.Host,
+	topo topology.Topology,
+	topoState *topology.StateSnapshot) error {
+	shard.bootstrapLock.Lock()
+	// check whether shard is already bootstrapping
+	if shard.bootstrapState > bootstrap.BootstrapNotStarted && shard.bootstrapState < bootstrap.Bootstrapped {
+		shard.bootstrapLock.Unlock()
+		return bootstrap.ErrTableShardIsBootstrapping
+	}
+	shard.bootstrapState = bootstrap.Bootstrapping
+	shard.bootstrapLock.Unlock()
+
+	success := false
+	defer func() {
+		shard.bootstrapLock.Lock()
+		if success {
+			shard.bootstrapState = bootstrap.Bootstrapped
+		} else {
+			shard.bootstrapState = bootstrap.BootstrapNotStarted
+		}
+		shard.bootstrapLock.Unlock()
+	}()
+
+	// find peer node for copy data
+	peerNode := shard.findBootstrapSource(origin, topo, topoState)
+	var dataStreamErr error
+	borrowErr := peerSource.BorrowConnection(peerNode.ID(), func(nodeClient rpc.PeerDataNodeClient) {
```

Comment:
so looks like one shard copy will use one connection, possible to distribute to multiple connection?

Reply:
grpc connection is ok to be shared and concurrently accessed. actually the number of streams is recommended to not be less than 100.
actually we will have a tighter limit on the number of concurrent raw data fetch streams (per vector party), enforced by the worker pool capacity.